### ISSUE TITLE
Mocking async functions

### ISSUE THREAD

**Issue by anton-johansson** [dvabuzyarov/moq.ts#578]
After messing around a bit, I found that this is a way of mocking async functions:

```ts
    /* eslint-disable @typescript-eslint/promise-function-async */
    const sessionManager = new Mock<SessionManager>()
        .setup(instance => instance.createSession(It.IsAny(), It.IsAny())).returns(Promise.resolve("my-session-id"))
        .object();
    /* eslint-enable @typescript-eslint/promise-function-async */
```

Some thoughts/ideas:
* Add documentation about this, because it seems like a fairly common case to mock async functions.
* Is it possible to do something a little more cleaner? For example, I have to disable this eslint warning, beacuse my function obviously returns a `Promise`, and my eslint ruleset wants this to be an async function. I have no idea how this could be improved though... :joy:

**Comment by dvabuzyarov** [comment]
Hello, I will take a look into this on the weekend.

**Comment by dvabuzyarov** [comment]
@anton-johansson Thanks for the ticket. I've implemented returnsAsync/throwsAsync methods.
ESLint rule could be fixed with using async keyword in setups, but it will break auto-mocking feature. So if you want to use it you need to do additional setup of moq.ts library.

```typescript
Mock.options = {
    injectorConfig: new EqualMatchingInjectorConfig([], [
        {
            provide: ReturnsAsyncPresetFactory,
            useClass: MimicsResolvedAsyncPresetFactory,
            deps: [RootMockProvider, Presets, ResolvedPromiseFactory]
        },
        {
            provide: ThrowsAsyncPresetFactory,
            useClass: MimicsRejectedAsyncPresetFactory,
            deps: [RootMockProvider, Presets, RejectedPromiseFactory]
        },
    ])
};
``` 
You can try it out with 7.4.0 version.

**Comment by anton-johansson** [comment]
Looks great! And `throwsAsync` as well, nice work. :)

I tried it, and while it does work, I found some strange stuff. Previously, I did this:
```ts
        .setup(x => x.getUser(Any())).returns(Promise.resolve(null))
        .setup(x => x.getUser("some-example@example.com")).returns(Promise.resolve(DATABASE_USER))
```

This effecitvely means that anything that isn't `some-example@example.com` will be returned as `null`, which have worked great. Now, when I try the same with `returnsAsync`:
```ts
        .setup(async x => x.getUser(Any())).returnsAsync(null)
        .setup(async x => x.getUser("some-example@example.com")).returnsAsync(DATABASE_USER)
```

... it seems like `DATABASE_USER` is always returned. Or `null` is returned if I swap their places. Just as if it is always the latest mock that is triggered.

This seems to be more related to the `async` keyword. If I remove the `async`-keyword and my custom `Mock.options`, it also works great (but I get the ESLint errors). Maybe I'm specifying my options incorrectly? I had a hard time knowing which actual imports needed to be used. I did this:

```ts
import {EqualMatchingInjectorConfig, Mock, ResolvedPromiseFactory, RejectedPromiseFactory} from "moq.ts";
import {ReturnsAsyncPresetFactory, ThrowsAsyncPresetFactory, MimicsResolvedAsyncPresetFactory, MimicsRejectedAsyncPresetFactory, RootMockProvider, Presets} from "moq.ts/internal";
```

Is this correct?

**Comment by dvabuzyarov** [comment]
@anton-johansson Thanks for the bug. Try out v7.4.1.

**Comment by anton-johansson** [comment]
You're a diamond, @dvabuzyarov! This works like a charm.

I just have one question regarding the imports of `ReturnsAsyncPresetFactory`, etc.

I get an ESLint error about duplicate imports. I'm not sure why. I'm doing this:
```
/* eslint-disable import/no-duplicates */
import {IMock, It, Mock, ResolvedPromiseFactory, RejectedPromiseFactory, EqualMatchingInjectorConfig} from "moq.ts";
import {ReturnsAsyncPresetFactory, ThrowsAsyncPresetFactory, MimicsResolvedAsyncPresetFactory, MimicsRejectedAsyncPresetFactory, RootMockProvider, Presets} from "moq.ts/internal";
/* eslint-enable import/no-duplicates */
```

Is it correct that I need to import from two different ones? Or am I doing something wrong?

**Comment by anton-johansson** [comment]
By the way, is it possible to do the same for verify, somehow?

Here's an example where I have to disable ESLint errors for verifies:
```ts
    /* eslint-disable @typescript-eslint/promise-function-async */
    redis.verify(instance => instance.hmset("session:newsessionid", is({body: BODY, userId: 666})));
    redis.verify(instance => instance.expire("session:newsessionid", 604800));
    /* eslint-enable @typescript-eslint/promise-function-async */
```

**Comment by dvabuzyarov** [comment]
@anton-johansson can you give more information on the issues? 
I tried the imports and there is no complaining. The imports it self look good.

```typescript
 it("verifies invocation", async () => {
        interface IUserManager {
            getUser(name: string): Promise<string>;
        }

        const injectorConfig = new EqualMatchingInjectorConfig([], [
            {
                provide: ReturnsAsyncPresetFactory,
                useClass: MimicsResolvedAsyncPresetFactory,
                deps: [RootMockProvider, Presets, ResolvedPromiseFactory]
            },
            {
                provide: ThrowsAsyncPresetFactory,
                useClass: MimicsRejectedAsyncPresetFactory,
                deps: [RootMockProvider, Presets, RejectedPromiseFactory]
            },
        ]);

        const value = "user";
        const name = "some-example@example.com";

        const managerMock = new Mock<IUserManager>({injectorConfig})
            .setup(async x => x.getUser(It.IsAny()))
            .returnsAsync(null)
            .setup(async x => x.getUser(name))
            .returnsAsync(value);

        const actual = await managerMock.object().getUser(name);

        expect(actual).toBe(value);
        managerMock.verify(async instance => instance.getUser(name));
    });
```

if you are using async in setup, then you can use it in verify. It works on my side. If I missed something, please let me know.

**Comment by anton-johansson** [comment]
Okay, lets keep the two issues separate.

1. ESLint gives me an error when I'm using both `from "moq.ts"` and `from "moq.ts/internal"`. I'm not sure why. It's not a super-big deal, because I've written a helper method to set up my mockings that I use everywhere, so I only need to disable the linting warning once. The error is saying:

```
'/home/anton/projects/my-project/node_modules/moq.ts/bundles/moq.ts.umd.js' imported multiple times. (import/no-duplicates) XO(import/no-duplicates)
```

2. Adding `async` in verifications simply does not work for me. Here is my verification that works:

```ts
/* eslint-disable-next-line @typescript-eslint/promise-function-async */
emailService.verify(instance => instance.send("email@example.com", "Reset password", "my-mocked-email-body"));
```

... now, if I change this to:
```ts
emailService.verify(async instance => instance.send("email@example.com", "Reset password", "my-mocked-email-body"));
```

... I get the following error message back:

```
Getter of 'then' of instance.send('email@example.com', 'Reset password', 'my-mocked-email-body') should be called once, but was called 0 time(s)
```

It looks just as if it gets confused by the return value being a `Promise`.

**Comment by dvabuzyarov** [comment]
@anton-johansson regarding verifying issue. It looks like you have not read "then" property of returned promise. Let me explain my assumption: there is the auto-making feature that allows to automatically create a chain of mocked objects. Consider this case:
```typescript
interface IChild {
    get(): string;
}

interface IRoot {
    child: IChild;
}

const value = "value";

const child = new Mock<IChild>()
    .setup(instance => instance.get())
    .returns(value)
    .object();

const root = new Mock<IRoot>()
    .setup(instance => instance.child)
    .returns(child)
    .object();

const actual = root.child.get();

expect(actual).toBe(value);
```

when we use async keyword in a setup section we are constructing a complex expression that is seen as 2 interactions: 
* invoking of the original function
* async feature reads "then" property of returned object

so we can rewrite our case in this way
```typescript
        interface IUserManager {
            getUser(name: string): Promise<string>;
        }

const value = "value";

const promiseMock = new Mock<Promise<string>>()
    .setup(instance => instance.get())
    .returns(value);

  const manager = new Mock<IUserManager>({injectorConfig})
      .setup(x => x.getUser(It.IsAny()))
      .returns(promiseMock.object())
      .object();

const actual = manager.getUser("name");

expect(actual).toBe(promiseMock.object()); // is true
promiseMock.verify(instance => instance.then); // is false

// or if we add await keyword in the action

const actual = await manager.getUser("name");

expect(actual).toBe(promiseMock.object()); // is true
promiseMock.verify(instance => instance.then); // is true!
```

so when you verify 
```typescript
emailService.verify(async instance => instance.send("email@example.com", "Reset password", "my-mocked-email-body"));
```
it verifies 2 interactions at once: 
* invocation of the send method with provided parameters
* reading "then" property of returned object

I can dig into disabling the auto-mocking feature and create a config for that, that will simplify of using async keyword.

The best solution would be ignoring the second interaction in case of applying async keyword, but at the moment I can't see a clear and robust way to figure it out in the express reflector.

[here is a good explanation of it](https://stackoverflow.com/questions/38508420/how-to-know-if-a-function-is-async)

**Comment by dvabuzyarov** [comment]
@anton-johansson about ESLint issue I can see now the root cause. I am using two packages to split up public and private API. Technically the entire API is public and the only difference d.ts files (public_api.d.ts and internal.d.ts) and the auxiliary internal package is using the same javascript files. ESLint can see it and complains. At the moment this approach is my best options.

If you have any ideas how to split up public and private API and satisfy ESLint rule at once, please share it with me.

**Comment by anton-johansson** [comment]
**Regarding async verifications:** Thanks a lot for the very detailed explanation. I **think** I undersatnd. The class function that I'm testing looks like this:
```ts
public async createResetPasswordToken(emailAddress: string) {
    const token = await this.userDao.createResetPasswordToken(emailAddress);
    const body = renderResetPasswordEmailBody(token);
    await this.emailService.send(emailAddress, "Reset password", body);
}
```

Now, in my tests, I've mocked both `createResetPasswordToken` and `renderResetPasswordEmailBody`, and I simply want to verify that `send` has been called with the correct arguments. I don't really care for the `then` call (altough, I think it should be called internally/automatically by `await`)?

Here's my test:

```ts
test("reset password, success", async () => {
    await service.createResetPasswordToken("email@example.com");

    emailService.verify(async instance => instance.send("email@example.com", "Reset password", "my mocked body: abc123"));
});
```

It would be great if I could just do this, as I'm only interested in verifying that the function has been called. Whether the promise succeeded or rejected isn't really of importance here, any potential error will make its way up to my generic error handling.


**Regarding ESLint and multiple imports:** Okay, I see. Unfortunately, I have no good input for you. I'm relatively new to TypeScript and JavaScript in general (coming form the Java world). But as I mentioned, it's perfectly fine for me, as I've abstracted away this logic into my own module, so I only have to ignore the ESLint error once. :)

**Comment by dvabuzyarov** [comment]
@anton-johansson Thanks for so detailed input. So at the end I found an issue with verification that is relative to the auto-mocking feature. This issue make the whole feature incomplete, so I am going to disable it. The disabling will fix all current issues with support for async/await keywords (I do hope ;-)).

**Comment by anton-johansson** [comment]
Cool! I'm perfectly fine with an opt-out feature, if that's best. However, if it causes various problems, maybe an opt-in is better for now.

**Comment by anton-johansson** [comment]
Okay, so I tried `8.0.0` and it seems to be causing some other troubles with my mocks:
```
TypeError: Function.prototype.apply was called on undefined, which is a undefined and not a function
```

I seem to get it as soon as a mocked function is called.

**Comment by dvabuzyarov** [comment]
@anton-johansson hmm... at the moment I strangled to reproduce the issue. May I ask you to give me a reproducible example?

**Comment by dvabuzyarov** [comment]
@anton-johansson or if it is acceptable for you we can do a pair session and sort it on your side.

**Comment by dvabuzyarov** [comment]
@anton-johansson Hello, do you still have the issue?

**Comment by anton-johansson** [comment]
Okay, so while preparing a new repository with some sample code, I noticed that it's related to my `Mock.options` override to handle `async` properly (so that I can keep using my ESLint error for async arrow functions).

Here is my sample code:
https://github.com/anton-johansson/bug-reports/tree/dvabuzyarov/moq.ts%2F578%2F808741580

If you remove the section that sets the `Mock.options`, tests will pass.

**However**, it turns out that I don't need toe `Mock.options` override anymore. It seems like `async` works out of the box now, and so by removing `Mock.options`, all my problems seem to be solved.

I'm therefore closing this issue! If you want to investigate the above issue with `Mock.options` and the `async` ones, the repository will be open. Thanks a lot for the quick responses and good explanations, @dvabuzyarov!

**PS:** I wrote this message a couple of days ago, but had forgotten to press "Comment" -.-
